### ISSUE TITLE
bam file realignment

### ISSUE THREAD

**Issue by razZ0r** [chapmanb/bcbio-nextgen#1189]
I have already aligned bam files from different sources, and I'd like to realign and process them in a uniform way. GATK has some [recommendations](http://gatkforums.broadinstitute.org/gatk/discussion/2908/howto-revert-a-bam-file-to-fastq-format) on how to do this. Step 1 is to shuffle the reads in the bam file. If I set `aligner: bwa` will the bcbio pipline do this shuffling, when converting to fastq? Is it even needed here? Can't seem to figure it out.

Also, bamtofastq starts dumping sequences on stdout. Is this normal? I did not wait for the jobs to finish, so it might write stuff to a file in the end, but it seemed strange.

<pre><code>
[2016-01-15T11:20Z] scurbat: Resource requests: bwa, sambamba, samtools; memory: 2.00, 2.00; cores: 16, 16, 16
[2016-01-15T11:20Z] scurbat: Configuring 1 jobs to run, using 8 cores each with 16.1g of memory reserved for each job
[2016-01-15T11:21Z] scurbat: Timing: organize samples
[2016-01-15T11:21Z] scurbat: ipython: organize_samples
[2016-01-15T11:24Z] node7: Using input YAML configuration: /data/t-tcga/esebesty/bcbio_run/single-pair-test/config/single-pair-test.yaml
[2016-01-15T11:25Z] node7: Checking sample YAML configuration: /data/t-tcga/esebesty/bcbio_run/single-pair-test/config/single-pair-test.yaml
[2016-01-15T11:25Z] node7: Testing minimum versions of installed programs
[2016-01-15T11:21Z] scurbat: Timing: alignment preparation
[2016-01-15T11:21Z] scurbat: ipython: prep_align_inputs
[2016-01-15T11:25Z] node7: BAM to bgzipped fastq
[2016-01-15T11:25Z] node7: @C2GCRACXX131008:3:1205:14570:83744/1
[2016-01-15T11:25Z] node7: AGGGTTAGGGTTAGGGTTAGGGTTAGGGAGATCGGAAGAGCACACGTCTGAACTACAGTCACAAATTGCCCTCTCGTATGCCGTCTTCTCCTTGCTAAAAT
[2016-01-15T11:25Z] node7: +
[2016-01-15T11:25Z] node7: ?<<?=<AACB?:;AB@:9=@AAB@<@?71.1098?25=6<*0>><7=-9:5-+><1>BB;B68:<A<C#################################
[2016-01-15T11:25Z] node7: @C2GCRACXX131008:3:1205:14570:83744/2
[2016-01-15T11:25Z] node7: CCCTAACCCTAACCCTAACCCTAACCCTAGATCGGAAGAGCGTCGTGTAGGGAACAGTGTAGATCTCGGCGGTCGCCGCGTCATTCAAAAAAATTCCGATA
</code></pre>

**Comment by chapmanb** [comment]
Endre;
We do an equivalent shuffle/collate and export step to the GATK recommendations but use `bamtofastq` from biobambam (https://github.com/gt1/biobambam2) which handles both collating and exporting instead of needing to pipe between two tools.

We don't expect reads to be output to stdout. Looking at the code, this might happen if bcbio doesn't correctly determine if the file is paired. It looks at the first 50000 lines of the bam file and checks if any are paired, otherwise assumes everything is a singleton. Is it possible the input is some kind of mix of paired and singleton, or something else strange is going on? If you can pass along a subset of the the BAM file to me privately by e-mail happy to have a look.

**Comment by chapmanb** [comment]
Endre;
Thank you for passing the file along. After failing to reproduce the issue with your input file I came across the likely issue. I believe this is a manifestation of the issue in #1188, bcbio is not finding `sambamba` because it expects the tool directory on your path. The code for checking was silently failing in these cases due to trying to work around issues with reading from the head of a sambamba pipe. The silent failure would result in the file getting classified as single end, leading to the problems you saw.

So the fix from your side is to put bcbio's tooldir in your PATH, and then it should correctly detect these as paired end. Practically I've also fixed bcbio to fail in these cases with a correct error message instead of having this confusing result.

Thanks much for the feedback and for helping debug this. Hope this gets you running cleanly with bcbio now.
